utils/s2n_safety.h

```diff
@@ -47,6 +47,38 @@ bool s2n_in_test();
 /* Returns true if a and b are equal, in constant time */
 bool s2n_constant_time_equals(const uint8_t* a, const uint8_t* b, const uint32_t len);
 
+/* Returns true if a and b are equal. Execution time may depend on len */
+/* but not on the value of the data denoted by a or b                  */
+/* Note that if len == 0, then returns true                            */
+bool s2n_constant_time_equals_partial(const uint8_t* const a,
+                                      const uint8_t* const b,
+                                      const uint32_t len)
```

Comment:
The "partial" version _can_ be called from other modules if those modules are proven to meet the pre-condition, so it needs to be exported. The "total" version is for modules that haven't been verified to meet the pre-conditions and/or you just don't trust.

Reply:
I'll be talking about this at CryptoCon in September. You coming?

Reply:
I really really want to be at CryptoCon, but may have to be on the East Coast for work that week.  Optimistic we can work it out.

Reply:
What's to stop someone from writing a new module and calling partial without the pre-conditions? If a function is declared in a header, one with safety in its name, folks are going to fairly assume that it's safe to call. But they could be generating UAF and segfault bugs. That can't be right.

Reply:
You're right. I really need a way of saying "this function is private and internal to this library", but I don't think there's a way of expressing that in C.

Reply:
If you declare it static in s2n_safety.c , it can't be exported, we do this in a few other places. The compiler might even just inline it and there won't even be an internal symbol for it. Can force that behavior with inline too.

Reply:
Done